docs/source/model_design/programming_vivarium/debugging/index.rst

```diff
@@ -0,0 +1,190 @@
+we can print or manipulate the dataframes being worked with. The code 
+for this in the python debugger is the same as used in 
+a Jupyter notebook. By printing the dataframes being used, you 
+might discover that one of the dataframes is empty - and the 
+issue is more clear! 
+
+From there, we can look at the code generating the empty dataframe. 
+Maybe you see the error quickly. But if not, we'll probably 
+want to add a breakpoint. 
+
+To quit the python debugger, enter :code:`quit`. 
+
+Adding Breakpoints
+------------------
+
+So, we have identified the issue is an empty dataframe. Let's keep 
+debugging! The datframe in question is generated 
+in a function not in our stack trace. This means we'll have trouble 
+getting to it with the python debugger. We'll need to set a breakpoint. 
+
+A breakpoint is a line of code :code:`breakpoint()` that "breaks" 
+the code. So when this line is read, the model will drop you into a 
+python debugger at that point in the code. So in our example, we might 
+set a breakpoint in the function which creates the empty dataframe. 
+
+Once in the python debugger, we can use a similar process to the above. 
+Start by printing the dataframe. If it's empty, the issue is earlier 
+in the code. If the dataframe is still correct, the issue is later. 
+Either use up or step (:code:`s`) to move up or "step" down to the next 
+line. In this manner, you can navigate around the code in the debugger and 
+find which line is the issue. From there, you can investigate the line 
+and fix the bug. 
+
+If you need to use multiple breakpoints, you can also use the 
+command :code:`c` for continue, to move until the next breakpoint. 
+
+Using iPython
+-------------
+
+iPython is less commonly used for the type of debugging outlined 
+above. Instead, it is helpful for doing some basic checks on 
+outputs, or manipulation of outputs without opening a Jupyter 
+notebook. 
+
+First, you'll need to install iPython in your environment with 
+:code:`pip install iPython`. Then start an iPython session by 
+typing :code:`iPython` in the command line. 
+
+From here, you can use this the same as a Jupyter notebook, except 
+code can't be stored and you must enter and run lines one at a time. Some effects 
+of this are, for example, you have to run all :code:`import` lines every 
+time you start an iPython session, and you can't store custom 
+functions to use repeatedly easily. However, there are some use cases. 
+
+For example, if you want to print out an artifact key to check the 
+columns included. Jupyter notebook can do this, but you will need 
+to open a notebook, connect online, create a file, print the file, 
+and at the end of that, you'll need to end the Jupyter session before you 
+can run anything else from the open terminal. This is a lot of work to visualize 
+one dataframe! iPython can be run in the 
+terminal without creating unneeded files or worrying about connections. 
```

Comment:
Are you assuming you have the Artifact stored locally on your laptop? If not, won't you still have to connect to the cluster to see the data?